ddtrace/internal/_encoding.pyx

```diff
@@ -91,180 +91,223 @@ cdef class Packer(object):
                 if L > ITEM_LIMIT:
-                    PyErr_Format(ValueError, b"%.200s object is too large", Py_TYPE(o).tp_name)
-                rawval = o
+                    raise ValueError("unicode string is too large")
                 ret = msgpack_pack_raw(&self.pk, L)
                 if ret == 0:
-                    ret = msgpack_pack_raw_body(&self.pk, rawval, L)
-            elif PyUnicode_Check(o):
-                if self.encoding == NULL:
-                    ret = msgpack_pack_unicode(&self.pk, o, ITEM_LIMIT)
-                    if ret == -2:
-                        raise ValueError("unicode string is too large")
-                else:
-                    o = PyUnicode_AsEncodedString(o, self.encoding, self.unicode_errors)
-                    L = len(o)
-                    if L > ITEM_LIMIT:
-                        raise ValueError("unicode string is too large")
-                    ret = msgpack_pack_raw(&self.pk, L)
-                    if ret == 0:
-                        rawval = o
-                        ret = msgpack_pack_raw_body(&self.pk, rawval, L)
-            elif PyDict_CheckExact(o):
-                d = <dict>o
-                L = len(d)
-                if L > ITEM_LIMIT:
-                    raise ValueError("dict is too large")
-                ret = msgpack_pack_map(&self.pk, L)
-                if ret == 0:
-                    for k, v in d.items():
-                       ret = self._pack(k)
-                       if ret != 0: break
-                       ret = self._pack(v)
-                       if ret != 0: break
-            elif PyList_CheckExact(o):
-                # Expect a list of traces or a list of spans
-                L = len(o)
-                if L > ITEM_LIMIT:
-                    raise ValueError("list is too large")
+                    ret = msgpack_pack_raw_body(&self.pk, <char *> text, L)
+            return ret
 
-                ret = msgpack_pack_array(&self.pk, L)
-                if ret != 0:
-                    break
+        raise TypeError("Unhandled text type: %r" % type(text))
 
-                if L > 0 and PyList_CheckExact(o[0]):
-                    # List of lists of spans (a list of traces)
-                    for i in range(L):
-                        span = o[i]
-                        ret = self._pack(span)
-                        if ret != 0: break
-                else:
-                    # List of spans
-                    for i in range(L):
-                        ret = self._pack(o[i])
-                        if ret != 0: break
+    cdef inline int _pack_meta(self, object meta):
+        cdef Py_ssize_t L
+        cdef int ret
+        cdef dict d
 
-            elif isinstance(o, Span):
-                has_span_type = <bint>(o.span_type is not None)
-                has_meta = <bint>(len(o.meta) > 0)
-                has_metrics = <bint>(len(o.metrics) > 0)
+        if meta is None:
+            ret = msgpack_pack_nil(&self.pk)
```

Comment:
this is new?

so we'll get:

```python
{
  "meta": None
}
```

or 
```python
{
  "meta": {
    "key": "value"
  }
}
```

?

if so, this might actually cause issues with the agent.
this is what is gets unpacked into: https://github.com/DataDog/datadog-agent/blob/6265feee4338a9cadedf974efcb973f3f2c9ba3d/pkg/trace/pb/span.pb.go#L54

and Go is strict about typing for unpacking, so if that didn't match exactly it would throw an exception on decoding.

Reply:
wait, looking at `_pack_span` it seems that we already do a check for whether `meta` exists and is set, so do we even need this check/nil setting here?

L209

> has_meta = <bint> (len(span.meta) > 0)

Reply:
As far as this change goes I think the check could be dropped; with 0.5 we would probably need a `None` check that encodes an empty map

Reply:
> we would probably need a None check that encodes an empty map

Unless the agent properly handles `nil` we'll 100% need to take this approach instead.

Reply:
> > has_meta =  (len(span.meta) > 0)
It seems like this check doesn't actually check the case if meta exists, so what if we:
- Check if meta exists in `_pack_span()`
- Check if meta is set (i.e. `len(span.meta) > 0`) in `_pack_meta()`.

@brettlangdon  @P403n1x87 If `span.meta` is an empty map, what do we want to do?

Reply:
We discussed off of GitHub, we just need to remove the None check/nil writing here, that way `_pack_meta`/`_pack_metrics` will always either write a mapping or raise an exception.

Since we (today) only ever call `_pack_meta`/`_pack_metrics` if they are non-empty dicts, so the None check is also redundant.